### PR TITLE
Add Keadm init-beta/manifest generate

### PR DESCRIPTION
**Pull request by zhu733756**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Part of #3475 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
add keadm init-beta/manifests generate cmds
```

---------------------------------

## Introduction
```
make all WHAT=keadm

keadm init-beta" command install KubeEdge's master node (on the cloud) component by using a list of set flags like helm.
It checks if the Kubernetes Master are installed already,
If not installed, please install the Kubernetes first.

Usage:
  keadm init-beta [flags]

Examples:

keadm init-beta

- This command will render and install the Charts for Kubeedge cloud component

keadm init-beta --advertise-address=127.0.0.1 [--set cloudcore-tag=v1.9.0] --profile version=v1.9.0 -n kubeedge --kube-config=/root/.kube/config

  - kube-config is the absolute path of kubeconfig which used to secure connectivity between cloudcore and kube-apiserver


Flags:
      --advertise-address string    Use this key to set IPs in cloudcore's certificate SubAltNames field. eg: 10.10.102.78,10.10.102.79
      --cloudcore-image string      The whole image of the cloudcore, default is kubeedge/cloudcore:v1.9.0
      --cloudcore-tag string        The image tag of the cloudcore, default is v1.9.0
  -d, --dry-run                     Print the generated k8s resources on the stdout, not actual excute. Always use in debug mode
      --external-helm-root string    External helm root path
  -f, --files string                Allow appending file directories of k8s resources to keadm, separated by commas
      --force                       Forced installing the cloud components.
  -h, --help                        help for init-beta
      --iptablesmgr-image string    The whole image of the iptables manager, default is kubeedge/cloudcore:v1.9.0
      --iptablesmgr-tag string      The image tag of the iptables manager, default is v1.9.0
      --kube-config string          Use this key to set kube-config path, eg: $HOME/.kube/config (default "/root/.kube/config")
      --manifests string            Allow appending file directories of k8s resources to keadm, separated by commas
  -n, --namespace string            Namespace to install kubeedge cloud component, default is kubeedge
      --profile string              set profile on the command line (iptablesMgrMode=external or version=1.9.1)
      --set stringArray             set values on the command line (can specify multiple or separate values with commas: key1=val1,key2=val2)
```

```
keadm manifest generate

- This command will render and generate Kubernetes resources

keadm manifest generate --advertise-address=127.0.0.1 [--set cloudcore-tag=v1.9.0] --profile version=v1.9.0 -n kubeedge --kube-config=/root/.kube/config

  - kube-config is the absolute path of kubeconfig which used to secure connectivity between cloudcore and kube-apiserver


Flags:
      --advertise-address string    Use this key to set IPs in cloudcore's certificate SubAltNames field. eg: 10.10.102.78,10.10.102.79
      --cloudcore-image string      The whole image of the cloudcore, default is kubeedge/cloudcore:v1.9.0
      --cloudcore-tag string        The image tag of the cloudcore, default is v1.9.0
  -d, --dry-run                     Print the generated k8s resources on the stdout, not actual excute. Always use in debug mode
      --external-helm-root string   External helm root path.
  -f, --files string                Allow appending file directories of k8s resources to keadm, separated by commas
  -h, --help                        help for generate
      --iptablesmgr-image string    The whole image of the iptables manager, default is kubeedge/cloudcore:v1.9.0
      --iptablesmgr-tag string      The image tag of the iptables manager, default is v1.9.0
      --kube-config string          Use this key to set kube-config path, eg: $HOME/.kube/config (default "/root/.kube/config")
      --manifests string            Allow appending file directories of k8s resources to keadm, separated by commas
  -n, --namespace string            Namespace to install kubeedge cloud component, default is kubeedge
      --profile string              set profile on the command line (iptablesMgrMode=external or version=1.9.1)
      --set stringArray             set values on the command line (can specify multiple or separate values with commas: key1=val1,key2=val2)
      --skip-crds                  Skip installing/rendering the CRDs.
```

## How to use

1 Examples of `keadm init-beta`
```
_output/local/bin/keadm init-beta --advertise-address=192.168.88.6  -n kubeedge --kube-config=/root/.kube/config --force
_output/local/bin/keadm init-beta --set server.advertiseAddress={192.168.88.6} --set server.nodeName=allinone  -n kubeedge --kube-config=/root/.kube/config --force --external-helm-root=/root/go/src/github.com/edgemesh/build/helm --profile=edgemesh
```
2 Examples of `keadm manifests generate`:
```
_output/local/bin/keadm manifest generate --advertise-address=192.168.88.6  -n kubeedge --kube-config=/root/.kube/config [--skip-crds]

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -104,3 +104,5 @@ replace (
 	k8s.io/utils v0.0.0 => k8s.io/utils v0.21.4
 	sigs.k8s.io/apiserver-network-proxy/konnectivity-client => sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.22
 )
+
+replace google.golang.org/grpc => google.golang.org/grpc v1.26.0
```
Comment (gy95) [inline go.mod:108]: we don't need this line? there is already a `google.golang.org/grpc v1.38.0` in line 39

Comment (zhu733756) [reply go.mod:108]: This fixed grpc version are not compatible, see https://github.com/asim/go-micro/issues/2091

File: keadm/cmd/keadm/app/cmd/util/helminstaller.go [hunk 2]
```diff
@@ -126,11 +127,26 @@ func (cu *KubeCloudHelmInstTool) RunHelmInstall(baseHelmRoot string) error {
...
+	}
+
+	fmt.Printf("=========CHART DETAILS=======\n")
+	fmt.Printf("NAME: %s\n", release.Name)
+	if !release.Info.LastDeployed.IsZero() {
+		fmt.Printf("LAST DEPLOYED: %s\n", release.Info.LastDeployed.Format(time.ANSIC))
+	}
+	fmt.Printf("NAMESPACE: %s\n", release.Namespace)
+	fmt.Printf("STATUS: %s\n", release.Info.Status.String())
+	fmt.Printf("REVISION: %d\n", release.Version)
```
Comment (zhu733756) [inline keadm/cmd/keadm/app/cmd/util/helminstaller.go:148]: The above codes will print:
```
Kubernetes version verification passed, KubeEdge installation will start...
Cloudcore started
=========CHART DETAILS=======
NAME: cloudcore
LAST DEPLOYED: Fri Jan 14 11:01:47 2022
NAMESPACE: kubeedge
STATUS: deployed
REVISION: 1
```

File: keadm/cmd/keadm/app/cmd/util/helminstaller.go [hunk 3]
```diff
@@ -275,17 +290,42 @@ func (cu *KubeCloudHelmInstTool) runHelmManifest(r *Renderer, stdout io.Writer)
...
+	// let the os.Stdout not print the details
+	logFunc := func(format string, v ...interface{}) {}
 	if err := cfg.Init(cf, cu.Namespace, "", logFunc); err != nil {
-		return err
+		return nil, err
+	}
+
+	// a flag to confim the install/upgrade action
+	var performInstall bool
+	_, err := cfg.Releases.Last(r.componentName)
```
Comment (zhu733756) [inline keadm/cmd/keadm/app/cmd/util/helminstaller.go:307]: Make a quick request to check the release exists or not to do install or upgrade.

File: go.mod (hunk 1 above)
Comment (fisherxu) [reply go.mod:108]: How will it going if we move this line to L39?

Comment (zhu733756) [reply go.mod:108]: > How will it going if we move this line to L39?

```
github.com/kubeedge/kubeedge/edge/pkg/metamanager/metaserver/kubernetes/storage/sqlite/imitator imports
        k8s.io/apiserver/pkg/storage/etcd3 tested by
        k8s.io/apiserver/pkg/storage/etcd3.test imports
        go.etcd.io/etcd/integration imports
        go.etcd.io/etcd/proxy/grpcproxy imports
        google.golang.org/grpc/naming: module google.golang.org/grpc@latest found (v1.43.0), but does not contain package google.golang.org/grpc/naming
```

Comment (zhu733756) [reply go.mod:108]: I have compressed the commits that would be easier to review. :）

File: keadm/cmd/keadm/app/cmd/util/helminstaller.go [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+
+	release, err := cu.runHelmInstall(renderer)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s started\n", strings.ToTitle(renderer.componentName))
+
+	if release == nil {
+		return nil
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/util/helminstaller.go:138]: return error?

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		"Namespace to install kubeedge cloud component, default is kubeedge")
+
+	cmd.Flags().BoolVarP(&BetaInitOpts.DryRun, types.DryRun, "d", BetaInitOpts.DryRun,
+		"Print the generated k8s resources on the stdout, not actual excute. Always use in debug mode")
+
+	cmd.Flags().StringVar(&BetaInitOpts.CloudcoreImage, types.CloudcoreImage, BetaInitOpts.CloudcoreImage,
+		"The whole image of the cloudcore, default is kubeedge/cloudcore:v1.9.0")
+
+	cmd.Flags().StringVar(&BetaInitOpts.CloudcoreTag, types.CloudcoreTag, BetaInitOpts.CloudcoreTag,
+		"The image tag of the cloudcore, default is v1.9.0")
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/beta/init-beta.go:107]: CloudcoreImage is like `kubeedge/cloudcore`, without image tag,  and CloudcoreTag is like `v1.9.0`?

File: keadm/cmd/keadm/app/cmd/util/helminstaller.go (hunk 4 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/util/helminstaller.go:138]: will fix these

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		"Use this key to set kube-config path, eg: $HOME/.kube/config")
+
+	cmd.Flags().StringVar(&BetaInitOpts.Manifests, types.Manifests, BetaInitOpts.Manifests,
+		"Allow appending file directories of k8s resources to keadm, separated by commas")
+
+	cmd.Flags().StringVarP(&BetaInitOpts.Manifests, types.Files, "f", BetaInitOpts.Manifests,
+		"Allow appending file directories of k8s resources to keadm, separated by commas")
+
+	cmd.Flags().StringVarP(&BetaInitOpts.Namespace, types.Namespace, "n", BetaInitOpts.Namespace,
+		"Namespace to install kubeedge cloud component, default is kubeedge")
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/beta/init-beta.go]: This part has removed at the latest codes.

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			checkFlags := func(f *pflag.Flag) {
+				util.AddToolVals(f, flagVals)
+			}
+			cmd.Flags().VisitAll(checkFlags)
+			err := AddBetaInit2ToolsList(tools, flagVals, BetaInit)
+			if err != nil {
+				return err
+			}
+			return ExecuteBetaInit(tools)
```
[the remark this answers is not in the document]
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/beta/init-beta.go:70]: handle like this:
```
	if !cu.Force && cu.ExternalHelmRoot == "" {
		cloudCoreRunning, err := cu.IsKubeEdgeProcessRunning(KubeCloudBinaryName)
		if err != nil {
			return err
		}
		if cloudCoreRunning {
			return fmt.Errorf("CloudCore is already running on this node, please run reset to clean up first")
		}
	}

	err := cu.IsK8SComponentInstalled(cu.KubeConfig, cu.Master)
	if err != nil {
		return err
	}
```

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go (hunk 5 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/beta/init-beta.go:107]: yes, will fix

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/beta/init-beta.go]:
```
=> Namespace:  constants.SystemNamespace,
```

File: keadm/cmd/keadm/app/cmd/util/helminstaller.go (hunk 4 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/util/helminstaller.go:138]: Have handled before the code.
```
	release, err := cu.runHelmInstall(renderer)
	if err != nil {
		return err
	}

	fmt.Printf("%s started\n", strings.ToTitle(renderer.componentName))

	if release == nil {
		return nil
	}
```

Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/util/helminstaller.go:138]: These codes prevent the empty point and are only used for print.

Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/util/helminstaller.go]: fixed

**Comment by zhu733756** [comment]
/retest

https://github.com/zhu733756/kubeedge/actions/runs/1816355749

**Review by gy95 — COMMENTED** [review]
lgtm

**Comment by zhu733756** [comment]
> The license check failed, seems the old version of etcd has been vendored, we can use the `go mod graph | grep etcd` to check which pkg import the old version of etcd. If possible, to upgrade it.

Wohuo, let me see see.

**Comment by zhu733756** [comment]
> The license check failed, seems the old version of etcd has been vendored, we can use the `go mod graph | grep etcd` to check which pkg import the old version of etcd. If possible, to upgrade it.

Done, PTAL.

File: build/helm/charts/profiles/iptablesmgr.yaml [hunk 8]
```diff
@@ -0,0 +1,95 @@
+# Default values for kubeedge.
```
[the remark this answers is not in the document]
Comment (fisherxu) [reply build/helm/charts/profiles/iptablesmgr.yaml:1]: And for every release version, we all need to create a file?

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+		"The image repo of the cloudcore, default is kubeedge/cloudcore")
+
+	cmd.Flags().StringVar(&BetaInitOpts.CloudcoreTag, types.CloudcoreTag, BetaInitOpts.CloudcoreTag,
+		"The image tag of the cloudcore, default is v1.9.0")
+
+	cmd.Flags().StringVar(&BetaInitOpts.IptablesMgrImage, types.IptablesMgrImage, BetaInitOpts.IptablesMgrImage,
+		"The image repo of the iptables manager, default is kubeedge/iptables-manager")
+
+	cmd.Flags().StringVar(&BetaInitOpts.IptablesMgrTag, types.IptablesMgrTag, BetaInitOpts.IptablesMgrTag,
+		"The image tag of the iptables manager, default is v1.9.0")
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/beta/init-beta.go:113]: Can we get the default verison from the git tag as in `currentVersion := version.Get().String()`, ref: https://github.com/kubeedge/kubeedge/pull/3554/commits/01297e595bf75e0cbdfb3d69f45f0ffca5559e44#diff-80858c982a00cd3e4c8c624dce259974735ee88de28cd6b1df8ba2737c3a2f3eR116

File: build/helm/charts/profiles/iptablesmgr.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhu733756) [reply build/helm/charts/profiles/iptablesmgr.yaml:1]: > This file is no needed? Seems similar with `[build/helm/charts/profiles/version.yaml]`

A little different with `mode: "external"`, I will remove it.

> And for every release version, we all need to create a file?
 
No need if has no break changes to helm, just patch the image.

File: keadm/cmd/keadm/app/cmd/beta/init-beta.go (hunk 9 above)
Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/beta/init-beta.go:113]: Sure

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

File: keadm/cmd/keadm/app/cmd/util/helminstaller.go [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+	return nil
+}
+
+// RunHelmInstall renders the Charts with the given values, then installs the Charts to the cluster.
+func (cu *KubeCloudHelmInstTool) RunHelmInstall(baseHelmRoot string) error {
+	// --force would not care about whether the cloud components exist or not
+	// Also, if gives a external helm root, no need to check and verify. Because it is always not a cloudcore.
+	if !cu.Force && cu.ExternalHelmRoot == "" {
+		cloudCoreRunning, err := cu.IsKubeEdgeProcessRunning(KubeCloudBinaryName)
+		if err != nil {
```
Comment (gy95) [inline keadm/cmd/keadm/app/cmd/util/helminstaller.go:103]: This function seems to use `pid cloudcore 2>&1`  command to predict whether cloudcore is running or not. For helm install mode, we couldn't use this way. we could check pods to do it? @zhu733756

Comment (zhu733756) [reply keadm/cmd/keadm/app/cmd/util/helminstaller.go:103]: > his function seems to use `pid cloudcore 2>&1` command to predict whether cloudcore is running or not. For helm install mode, we couldn't use this way. we could check pods to do it? @zhu733756

Yeah, checking pods is the correcting style of the k8s, last time I see this function can cover it, so reuse it.  I wonder if the former deployment mode is binary mode, we can also use it to check.

**Pull request merged**